src/api/c/memory.cpp

```diff
@@ -1,26 +1,79 @@
 /*******************************************************
- * Copyright (c) 2014, ArrayFire
+ * Copyright (c) 2019, ArrayFire
  * All rights reserved.
  *
  * This file is distributed under 3-clause BSD license.
  * The complete license agreement can be obtained at:
  * http://arrayfire.com/licenses/BSD-3-Clause
  ********************************************************/
 
+#include <memoryapi.hpp>
+
 #include <Array.hpp>
 #include <backend.hpp>
 #include <common/err_common.hpp>
 #include <handle.hpp>
-#include <memory.hpp>
 #include <platform.hpp>
 #include <af/backend.h>
 #include <af/device.h>
 #include <af/dim4.hpp>
+#include <af/memory.h>
 #include <af/version.h>
-#include <cstring>
 
 using namespace detail;
 
+BufferInfo &getBufferInfo(const af_buffer_info pairHandle) {
+    return *(BufferInfo *)pairHandle;
+}
+
+af_buffer_info getHandle(BufferInfo &pair) {
+    BufferInfo *pairHandle;
+    pairHandle = &pair;
+    return (af_buffer_info)pairHandle;
+}
+
+af_err af_create_buffer_info(af_buffer_info *pairHandle, void *ptr,
+                             af_event event) {
+    try {
+        BufferInfo *pair = new BufferInfo({ptr, event});
+        *pairHandle      = getHandle(*pair);
+    }
+    CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+af_err af_release_buffer_info(af_buffer_info pairHandle) {
+    try {
+        /// NB: deleting a memory event pair does NOT free the associated memory
+        /// and does NOT delete the associated event.
+        delete (BufferInfo *)pairHandle;
```

Comment:
Is there a technical reason why this doesn't own its members?

Reply:
I suppose we could do this.

I'd thought that it seemed unsafe to immediately `memFree` the ptr and destroy the event if we released the `buffer_info` (some sort of queue interaction seemed scary), but maybe we could at least destroy the event considering that's what the old `MemoryEventPair` implementation did. I'll add it and see how it looks.

Reply:
I actually changed my mind, and I still think not deleting the `af_event` (and freeing the memory seems dangerous either way) is best. It's entirely possible one might want to grab the event from an `af_buffer_info` and use it somewhere else [this actually happens somewhere - I forget where], in which case they'd have to zero out (`af_buffer_info_set_event`) every `af_buffer_info` that the event had lived in before releasing them, or refrain from releasing them altogether until they're done with the `af_event`. Seems a lot more error-prone to me. Since one needs to call `af_create_event` externally to creating an `af_buffer_info`, I think the expectation that they also free the `af_event` is reasonable.

Reply:
Yeah, the alloc functions uses the pointer after the `buffer_info` object has been released. Makes sense because we don't need to keep track of the event once we wait on it.

In that case, I think we should consider renaming this function to avoid confusion with other af_release* functions. The release function reduce the reference counts of the other types and their underlying resources. If we change the meaning for this object it could cause inconsistencies in the API.

Suggested API:

```C++
/// \brief Disassociates the pointer from the \ref af_buffer_info object
///
/// Gets the pointer and disassociated it from the \ref af_buffer_info object. Deleting
/// the af_buffer_info object will not affect this pointer.
///
/// param[out] event The \ref pointer that will be disassociated. If NULL no pointer is
///                  returned and the data is NOT freed.
/// param[in] buf The target \ref af_buffer_info object
/// \returns AF_SUCCESS
af_err af_unlock_buffer_info_pointer(void** ptr, af_buffer_info buf);

/// \brief Disassociates the \ref af_event from the \ref af_buffer_info object
///
/// Gets the \ref af_event and disassociated it from the af_buffer_info object. Deleting
/// the af_buffer_info object will not affect this event.
///
/// param[out] event The \ref af_event that will be disassociated. If NULL no event is
///                   returned and the event is NOT freed
/// param[in] buf The target \ref af_buffer_info object
/// \returns AF_SUCCESS
af_err af_unlock_buffer_info_event(af_event* event, af_buffer_info buf);

/// \brief deletes the \ref af_buffer_info and the resources its tracking
///
///  Deletes the \ref af_buffer_info object and its tracked resources. If buffer still holds
/// the pointer,  that pointer is freed after its associated event has triggered.
///
/// \param[in] buf The af_buffer_info object that will be deleted
/// \returns AF_SUCCESS
af_err af_delete_buffer_info(af_buffer_info buf);
```

Reply:
This plan sounds reasonable. I'll add these functions to the C++ API as well.

In terms of implementation: unlocking the event is simple, but in terms of freeing the pointer, should that be accomplished with `nativeFree` directly? Can't see another way.